### PR TITLE
Eliminate direct access to size/strides of THTensor; replace them with std::vector

### PR DESCRIPTION
**Pull request by ezyang**
* THTensor now stores `sizes_` and `strides_` which is a `std::vector<int64_t>`
* Anywhere a "public" API function made use of a int64_t* of sizes, I opted to just finagle it out of the tensor using THTensor_getSizePtr rather than try to rewrite all of these sites to use ArrayRef. They should use ArrayRef eventually, but not yet.
* There are new utility functions for resizing sizes/strides in one go (THTensor_resizeDim), or replacing sizes and strides with completely new values (THTensor_setSizesAndStrides)
* Anywhere you said `t->size[n] = 0`, we now say `THTensor_setSizeAt(t, n, 0)`, ditto for strides
* Anywhere you said `t->size[n]`, we now say `t->size(n)` (coming soon: ditto for strides)

Previous review of just the `std::vector` change in #9518, but I'm planning to merge this all in one go.

Note for @gchanan: review from commit "ci" and after

### CODE REVIEW

**Comment by cpuhrsch** [comment]
Not sure if std::vector is a good longterm choice since size and stride is so short. It also can't be compiled into very efficient code. Ideally we'd template on that size and branch somewhere, at least that led to the best code within CPUApplyUtils.h. Otherwise SmartVector could be a good choice for now.

**Comment by ezyang** [comment]
@cpuhrsch Yeah, some sort of inline size/stride is in the roadmap, but it's not blocking for the merge so I haven't done it. When we add a change like this some benchmarking will be needed.

**Review by gchanan — APPROVED** [review]
I assume you are just trying to figure out where you need the to put the no-strict-overflow?  Also, why do you need it?

File: aten/src/TH/generic/THTensor.cpp [hunk 1]
```diff
@@ -229,8 +229,8 @@ THTensor *THTensor_(newView)(THTensor *tensor, THLongStorage *size)
   ptrdiff_t numel = THTensor_(nElement)(tensor);
   THTensor *self = THTensor_(new)();
   THLongStorage *inferred_size = THLongStorage_newInferSize(size, numel);
-  auto stride = THTensor_compute_stride(at::IntList(THTensor_getSizePtr(tensor), tensor->dim()),
-                                        at::IntList(THTensor_getStridePtr(tensor), tensor->dim()),
+  auto stride = THTensor_compute_stride(tensor->sizes(),
+                                        tensor->strides(),
```
[the remark this answers is not in the document]
Comment (ezyang) [reply aten/src/TH/generic/THTensor.cpp]: There were some direct decrements to `dim_` which I missed because I grepped for `dim_ =`

**Pull request closed without merging**